lib/fizzy/parser.cpp

```diff
@@ -35,208 +32,182 @@ std::tuple<bool, const uint8_t*> parseGlobalType(const uint8_t* pos)
-            }
-
-            case Instr::i32_const:
-            {
-                result.kind = ConstantExpression::Kind::Constant;
-                int32_t value;
-                std::tie(value, pos) = leb128s_decode<int32_t>(pos);
-                result.value.constant = static_cast<uint32_t>(value);
-                break;
-            }
-
-            case Instr::i64_const:
-            {
-                result.kind = ConstantExpression::Kind::Constant;
-                int64_t value;
-                std::tie(value, pos) = leb128s_decode<int64_t>(pos);
-                result.value.constant = static_cast<uint64_t>(value);
-                break;
-            }
-            }
-        } while (instr != Instr::end);
-
-        return {result, pos};
-    }
-};
+            result.kind = ConstantExpression::Kind::GlobalGet;
+            std::tie(result.value.global_index, pos) = leb128u_decode<uint32_t>(pos);
+            break;
+        }
+
+        case Instr::i32_const:
+        {
+            result.kind = ConstantExpression::Kind::Constant;
+            int32_t value;
+            std::tie(value, pos) = leb128s_decode<int32_t>(pos);
+            result.value.constant = static_cast<uint32_t>(value);
+            break;
+        }
+
+        case Instr::i64_const:
+        {
+            result.kind = ConstantExpression::Kind::Constant;
+            int64_t value;
+            std::tie(value, pos) = leb128s_decode<int64_t>(pos);
+            result.value.constant = static_cast<uint64_t>(value);
+            break;
+        }
+        }
+    } while (instr != Instr::end);
+
+    return {result, pos};
+}
 
 template <>
-struct parser<Global>
+inline parser_result<Global> parse(const uint8_t* pos)
 {
-    parser_result<Global> operator()(const uint8_t* pos)
-    {
-        Global result;
-        std::tie(result.is_mutable, pos) = parseGlobalType((pos));
-        std::tie(result.expression, pos) = parser<ConstantExpression>{}(pos);
+    Global result;
+    std::tie(result.is_mutable, pos) = parseGlobalType((pos));
+    std::tie(result.expression, pos) = parse_constant_expression(pos);
 
-        return {result, pos};
-    }
-};
+    return {result, pos};
+}
 
 template <>
-struct parser<Table>
+inline parser_result<Table> parse(const uint8_t* pos)
 {
-    parser_result<Table> operator()(const uint8_t* pos)
-    {
-        const uint8_t kind = *pos++;
-        if (kind != FuncRef)
-            throw parser_error{"unexpected table elemtype: " + std::to_string(kind)};
+    const uint8_t kind = *pos++;
+    if (kind != FuncRef)
+        throw parser_error{"unexpected table elemtype: " + std::to_string(kind)};
 
-        Limits limits;
-        std::tie(limits, pos) = parser<Limits>{}(pos);
+    Limits limits;
+    std::tie(limits, pos) = parse_limits(pos);
 
-        return {{limits}, pos};
-    }
-};
+    return {{limits}, pos};
+}
 
-template <>
-struct parser<std::string>
+inline parser_result<std::string> parse_string(const uint8_t* pos)
```

Comment:
This is not a template specialization as it is never used in `parse_vec()`. But we can make it so if easier to read.